### PR TITLE
Fix window spawning triggering ButtonInput<KeyCode>::just_pressed/just_released

### LINKED ISSUE #12273: Opening a Window on `keyboard_input.just_pressed` makes it trigger again
## Bevy version

`0.13`

## \[Optional\] Relevant system information

`Windows 11`

## What you did

I made a system similar to the one in the `keyboard_input` example. But on the check whether the key `A` was just pressed, I then spawn a window.

```rs
fn spawn_debug_window(input: Res<ButtonInput<KeyCode>>, mut commands: Commands) {
    if input.just_pressed(KeyCode::F1) {
        info!("just pressed");
        commands.spawn(Window {
            title: "Debug".to_owned(),
            ..Default::default()
        });
    }
}
```

## What went wrong

It spams opening new windows, the `just_pressed` is held for multiple frames.
I think this happens because it opens a new window, thus the focus is on the new window and the press counts as a new press?

```
2024-03-03T11:54:46.916247Z  INFO my_project::debug: just pressed
2024-03-03T11:54:46.930608Z  INFO bevy_winit::system: Creating new window "Debug" (25v4)
2024-03-03T11:54:46.967434Z  INFO my_project::debug: just pressed
2024-03-03T11:54:46.974474Z  INFO bevy_winit::system: Creating new window "Debug" (27v1)
2024-03-03T11:54:47.014095Z  INFO my_project::debug: just pressed
2024-03-03T11:54:47.036982Z  INFO bevy_winit::system: Creating new window "Debug" (29v1)
2024-03-03T11:54:47.078519Z  INFO my_project::debug: just pressed
2024-03-03T11:54:47.119463Z  INFO bevy_winit::system: Creating new window "Debug" (31v1)
```

### PR DESCRIPTION
**Pull request by SpecificProtagonist**
# Objective

Fix #12273

## Solution

– Only emit `KeyboardFocusLost` when the keyboard focus is lost
– ignore synthetic key releases too, not just key presses (as they're already covered by `KeyboardFocusLost`)

---

### Fixed

- Don't trigger `ButtonInput<KeyCode>::just_pressed`/`just_released` when spawning a window/focus moving between Bevy windows

### CODE REVIEW

File: crates/bevy_input/src/keyboard.rs [hunk 1]
```diff
@@ -117,13 +117,34 @@ pub fn keyboard_input_system(
...
-            key_code, state, ..
-        } = event;
-        match state {
-            ButtonState::Pressed => key_input.press(*key_code),
-            ButtonState::Released => key_input.release(*key_code),
+
+    // When a window loses focus, we receive a synthetic key release event for all held keys;
+    // when a window gains focus, we receive a synthetic key down event for all held keys.
+    // As a result, when a new window is spawned, we would detect a key up/key down pair for
+    // all held keys even though the user didn't do anything.
```
Comment (alice-i-cecile) [inline crates/bevy_input/src/keyboard.rs:124]: We should link to the associated winit issue here.

**Comment by alice-i-cecile** [comment]
Does this fix obsolete https://github.com/bevyengine/bevy/pull/12367 ?

File: crates/bevy_input/src/keyboard.rs (hunk 1 above)
Comment (alice-i-cecile) [reply crates/bevy_input/src/keyboard.rs:124]: Yes exactly :)

**Review by alice-i-cecile — CHANGES_REQUESTED** [review]
On board with this behavior, but I'd like to avoid accumulating tech debt here. Can you please factor out the new filtering code into a function and add some tests for it? :)

The other question I have is: "why should this be in `bevy_input`, rather than filtering events at the very top-level in `bevy_winit`? That will save us work, and more importantly, ensure our data doesn't desynchronize.

**Comment by alice-i-cecile** [comment]
> These events could be useful when reading KeyboardInput events since they carry window information. I'm not confident enough to assert that this isn't useful.

@UkoeHB @aevyrie @mockersf any strong opinions here? I'm fine to leave this as a more scoped fix for now, especially with the logic extracted and tested.

**Comment by shanecelis** [comment]
Hi, I'm on macOS 14.3.1 x86 exercising the `multiple_windows` example with this system:

```rust
fn keyboard_input_system(keyboard_input: Res<ButtonInput<KeyCode>>) {

    if keyboard_input.just_pressed(KeyCode::KeyA) {
        info!("'A' just pressed");
    }
    if keyboard_input.just_released(KeyCode::KeyA) {
        info!("'A' just released");
    }

    if keyboard_input.just_pressed(KeyCode::ShiftLeft) {
        info!("'Shift' just pressed");
    }
    if keyboard_input.just_released(KeyCode::ShiftLeft) {
        info!("'Shift' just released");
    }
}
```

The output is always pairs of press and release. I haven't been able to provoke two presses in a row (for the same key) or two releases in a row. However, if I press `Shift` in window a, switch focus to window b then release, there is no release message. Here's an annotated log.

```
// Press Shift.
2024-03-08T04:37:18.330416Z  INFO multiple_windows: 'Shift' just pressed
// Switch focus to window b. Release Shift. No log of a release.

// Press Shift. No log.
// Release Shift while focused on window b then it will finally emit a release message.
2024-03-08T04:37:26.599310Z  INFO multiple_windows: 'Shift' just released
```

Also if `Shift` is held down from outside the app, then focus is changed to the app, `Shift` is released, no release message is emitted.

If one presses `Shift`, changes focus to window b, changes focus back to window a, and releases, then the press and release messages are emitted as expected.

Exercising some of the same scenarios with the `A` key does not have some of this funny behavior. In fact you can press `A` switch to another app and when you release `A` in another app, and the release message is emitted by our app.

I hope this helps explain some of the behavior on macOS. Let me know if there are specific scenarios you'd like to see exercised.

**Comment by shanecelis** [comment]
Sure thing. Pressing `A` while not focused, switching to the app, then releasing `A` emits no logs. Pressing and releasing `A` then works as expected.

Spawning another window works fine for the `A` key, but behaves the same way for the `Shift` key as described in the annotated log. That is a release and press event for `Shift` are omitted.

**Comment by richchurcher** [comment]
@SpecificProtagonist sorry to be a pain, any chance you could update your branch to make it a tad easier to review? Be nice to get this one into 0.15 :slightly_smiling_face:

**Review by richchurcher — APPROVED** [review]

**Comment by richchurcher** [comment]
I note that this and #14379 are doing related things, though I'm not quite sure of the subtleties around whether one makes the other redundant :thinking:

**Review by alice-i-cecile — APPROVED** [review]

**Comment by Azorlogh** [comment]
I'm still receiving KeyboadFocusLost intermittently when switching between bevy windows :confused: I think this happens because sometimes the app updates after one window lost focus, but just before the other window gained focus.  
This sounds like a really tricky case to handle properly :grimacing: 

https://github.com/user-attachments/assets/5cacea5e-8f37-40ee-aba6-09d682d596b3

**Pull request merged**